src/languages/taggerscript.js

```diff
@@ -22,7 +22,7 @@ export default function(hljs) {
 
   const FUNCTION = {
     className: 'keyword',
-    begin: /\$(?!noop)[a-zA-Z][_a-zA-Z0-9]*/,
+    begin: /\$(?!noop)[_a-zA-Z0-9]*/,
```

Comment:
Can a function be a single character, is `_` by itself a valid function name?

Reply:
Class name looks wrong here (I know you didn't change it).  Probably should be `title.function`?  Would you mind updating?

Does the `(` always immediately follow the function name?

Reply:
Yes to all of the above, strictly speaking it would be `/\$(?!noop)[_a-zA-Z0-9]+/` actually. But for syntax highlighting the above works better, because an unescaped $ always starts a function, so highighting `$` already works well for this purpose.

Reply:
>  it would be /\$(?!noop)[_a-zA-Z0-9]+/ actually

We should go with that then.  We need to be as strict as is reasonable as this is also used for auto-detection.

> But for syntax highlighting the above works better, because an unescaped $ always starts a functio

I don't follow.

If the `(` is always immediate then we should just change this to a matcher without the `end` to be more precise.

```
match: /\$(?!noop)[_a-zA-Z0-9]+(?\s*\(=)/
```

If whitespace isn't allowed the \s* could be removed.

Reply:
Regarding the name, that was changed here:

https://github.com/highlightjs/highlight.js/commit/0c8940069e69395036a48c51371b85699f04da75#diff-976da83ed3c74e03a20bb75c0361c839f1c917c3c426d7145781ec508dbab6eb

> Things like `$if`, `$ne`, `$eq` look more like keywords in a
  conventional language, certainly not as function names being
  defined by user

I tend to agree on this. The language does not allow you to define your own functions. But not sure really how this is all used or turns out for formatting. TextMate grammars have a class specifically for "functions provided by the framework/library", something like this would probably fit if that exists.

Reply:
Ok. Wouldn't `match: /\$(?!noop)[_a-zA-Z0-9]+` be sufficient? Why the `(?=\()` isn't that captured by the `end: /\(/` already?

About being explicit I will then also change the current generic definition of escape from `\.` to a list of the actual characters, because this is actually a very limited set.

Reply:
> Why the (?=\() isn't that captured by the end: /\(/ already?

Is this a function:

```
$lookslikefunction but... HEY HOW ARE YOU #!*#@#&@#& 
how is the weather
(
```

If there is no middle to speak of then we don't need begin/end, we only need match.

Reply:
Is there no distinction between what `$_lossless` and `$if` are semantically in the language?

Reply:
I have made the definitions more explicit as suggested, including removing the end parameter and explicitly giving the possible escape sequences. I also fixed some issues with nesting of $noop()

> Is there no distinction between what $_lossless and $if are semantically in the language?

No, both are just internally defined functions that run with no or some parameters and return something. In the end that even is true for $noop(), just that this is a specific builtin that is otherwise treated as kind of a comment (and $noop() is a rather strong indicator that the language is matching).

There is no definitive list of all possible functions, as the language can be extended. But not with the language itself.